runtime/bcutil/ROMClassBuilder.cpp

```diff
@@ -248,6 +251,108 @@ ROMClassBuilder::buildROMClass(ROMClassCreationContext *context)
 	return result;
 }
 
+BuildResult
+ROMClassBuilder::handleAnonClassName(J9CfrClassFile *classfile)
+{
+	J9CfrConstantPoolInfo* constantPool = classfile->constantPool;
+	U_32 cpThisClassUTF8Slot = constantPool[classfile->thisClass].slot1;
+	U_32 originalStringLength = constantPool[cpThisClassUTF8Slot].slot1;
+	const char* originalStringBytes = (const char*)constantPool[cpThisClassUTF8Slot].bytes;
+	U_16 newUtfCPEntry = classfile->constantPoolCount - 1; /* last cpEntry is reserved for anonClassName utf */
+	U_32 i = 0;
+	BOOLEAN stringOrNASReferenceToClassName = FALSE;
+	J9CfrConstantPoolInfo *anonClassName = NULL;
+	BOOLEAN newCPEntry = TRUE;
+	UDATA newAnonClassNameLength = originalStringLength + 1 + ROM_ADDRESS_LENGTH + 1;
+	BuildResult result = OK;
+	PORT_ACCESS_FROM_PORT(_portLibrary);
+
+	/* Find if there are any Constant_String or CFR_CONSTANT_NameAndType references to the className.
+	 * If there are none we don't need to make a new cpEntry, we can overwrite the existing
+	 * one since the only reference to it is the classRef
+	 */
+	for (i = 0; i < classfile->constantPoolCount; i++) {
+		if ((CFR_CONSTANT_String == constantPool[i].tag)
+			|| (CFR_CONSTANT_NameAndType == constantPool[i].tag)
+		) {
+			if (cpThisClassUTF8Slot == constantPool[i].slot1) {
+				stringOrNASReferenceToClassName = TRUE;
+			}
+		}
+	}
+
+
+	if (!stringOrNASReferenceToClassName) {
+		/* do not need the new cpEntry so fix up classfile->constantPoolCount */
+		newCPEntry = FALSE;
+		classfile->constantPoolCount -= 1;
+	}
+
+	if (!newCPEntry) {
+		newUtfCPEntry = cpThisClassUTF8Slot;
+	}
+	anonClassName = &classfile->constantPool[newUtfCPEntry];
```

Comment:
Bit of a nitpick - as this is C++, the local declaration can occur on first use*.
`*` provided the code doesn't `goto` across its initialization.